File: src/main/java/leetcode/_22/generate_parentheses.go
package main

import "fmt"

func generateParenthesis(N int) []string {

	res := make([]string, 0)
	curr := make([]byte, N*2)

	var generate = func(int, int, int) {}
	generate = func(open int, close int, index int) {
		if open == N && close == N {
			res = append(res, string(curr))
		}

		if open < N {
			curr[index] = '('
			generate(open+1, close, index+1)
		}

		if close < open {
			curr[index] = ')'
			generate(open, close+1, index+1)
		}
	}

	generate(0, 0, 0)

	return res
}

// ***************** Testing *****************
func main() {
	fmt.Println(generateParenthesis(3))
}

File: src/main/java/leetcode/_22/GenerateParentheses.java
package leetcode._22;

import java.util.LinkedList;
import java.util.List;

public class GenerateParentheses {

    private final List<String> result = new LinkedList<>();
    private char [] current;

    public List<String> generateParenthesis(int N) {
        current = new char[N*2];
        generate(N, 0, 0, 0);
        return result;
    }

    private void generate(int N, int open, int close, int index) {
        if (open == N && close == N) {
            result.add(new String(current));
        }

        if (open < N) {
            current[index] = '(';
            generate(N, open + 1, close, index + 1);
        }
        if (close < open) {
            current[index] = ')';
            generate(N, open, close + 1, index + 1);
        }
    }
}
